source_files/edge/p_setup.cc

```diff
@@ -1535,226 +1350,174 @@ static void LoadXGL3Nodes(int lumpnum)
-					uint8_t *lp = psr->line;
-					while (*lp != 0 && *lp != 0x7D)
-						lp++; // find end of line or '}'
-					if (*lp == 0x7D)
-					{
-						numsectors++;
-						break; // end of block
-					}
-					if (psr->next >= psr->length)
-					{
-						numsectors++;
-						break; // end of lump
-					}
+		if (tok != epi::TOK_Ident)
+			I_Error("Malformed TEXTMAP lump.\n");
 
-					continue; // skip line
-				}
-			}
+		// check namespace
+		if (lex.Match("="))
+		{
+			lex.Next(section);
+			if (! lex.Match(";"))
+				I_Error("Malformed TEXTMAP lump: missing ';'\n");
+			continue;
 		}
-	}
 
-	sectors = new sector_t[numsectors];
-	Z_Clear(sectors, sector_t, numsectors);
+		if (! lex.Match("{"))
+			I_Error("Malformed TEXTMAP lump: missing '{'\n");
 
-	psr->next = 0; // restart from start of lump
-	while (1)
-	{
-		if (!GetNextBlock(psr, (uint8_t*)ident))
-			break;
-
-		if (epi::case_cmp(ident, "sector") == 0)
+		if (section == "sector")
 		{
 			float cz = 0.0f, fz = 0.0f;
 			int light = 160, type = 0, tag = 0;
 			char floor_tex[10];
 			char ceil_tex[10];
 			strcpy(floor_tex, "-");
 			strcpy(ceil_tex, "-");
-
-			// process sector block
-			while (1)
+			for (;;)
 			{
-				if (!GetNextAssign(psr, (uint8_t*)ident, (uint8_t*)val))
-				{
-					uint8_t *lp = psr->line;
-					while (*lp != 0 && *lp != 0x7D)
-						lp++; // find end of line or '}'
-					if (*lp == 0x7D)
-					{
-						break; // end of block
-					}
-					if (psr->next >= psr->length)
-					{
-						break; // end of lump
-					}
+				if (lex.Match("}"))
+					break;
 
-					continue; // skip line
-				}
-				// process assignment
-				if (epi::case_cmp(ident, "heightfloor") == 0)
-				{
-					fz = str2float(val, 0.0f);
-				}
-				else if (epi::case_cmp(ident, "heightceiling") == 0)
-				{
-					cz = str2float(val, 0.0f);
-				}
-				else if (epi::case_cmp(ident, "texturefloor") == 0)
-				{
-					Z_StrNCpy(floor_tex, val, 8);
-				}
-				else if (epi::case_cmp(ident, "textureceiling") == 0)
-				{
-					Z_StrNCpy(ceil_tex, val, 8);
-				}
-				else if (epi::case_cmp(ident, "lightlevel") == 0)
-				{
-					light = str2int(val, 160);
-				}
-				else if (epi::case_cmp(ident, "special") == 0)
-				{
-					type = str2int(val, 0);
-				}
-				else if (epi::case_cmp(ident, "id") == 0)
-				{
-					tag = str2int(val, 0);
-				}
+				std::string key;
+				std::string value;
+				epi::token_kind_e block_tok = lex.Next(key);
+
+				if (block_tok == epi::TOK_EOF)
+					I_Error("Malformed TEXTMAP lump: unclosed block\n");
+
+				if (block_tok != epi::TOK_Ident)
+					I_Error("Malformed TEXTMAP lump: missing key\n");
+
+				if (! lex.Match("="))
+					I_Error("Malformed TEXTMAP lump: missing '='\n");
+
+				block_tok = lex.Next(value);
+
+				if (block_tok == epi::TOK_EOF || block_tok == epi::TOK_ERROR || value == "}")
+					I_Error("Malformed TEXTMAP lump: missing value\n");
+
+				if (! lex.Match(";"))
+					I_Error("Malformed TEXTMAP lump: missing ';'\n");
+
+				if (key == "heightfloor")
```

Comment:
We lost the else if blocks here, which maaaaaybe would slow down loading a lot of sectors.  Frankly, probably negligible, might be a bit tidier as a switch statement.